File: Design-Patterns-Using-Go/Template/TemplatePatternData.go
package main

import "fmt"

// AddDataTemplate is the abstract class defining the template method.
type AddDataTemplate interface {
	addData()
	open()
	add()
	close()
}

// AddDataToFile is a concrete implementation of AddDataTemplate for adding data to a file.
type AddDataToFile struct{}

func (a *AddDataToFile) addData() {
	a.open()
	a.add()
	a.close()
}

func (a *AddDataToFile) open() {
	fmt.Println("Open file.")
}

func (a *AddDataToFile) add() {
	fmt.Println("Add data to file.")
}

func (a *AddDataToFile) close() {
	fmt.Println("Close file.")
}

// AddDataToDB is another concrete implementation of AddDataTemplate for adding data to a database.
type AddDataToDB struct{}

func (a *AddDataToDB) addData() {
	a.open()
	a.add()
	a.close()
}

func (a *AddDataToDB) open() {
	fmt.Println("Open Database.")
}

func (a *AddDataToDB) add() {
	fmt.Println("Add data to Database.")
}

func (a *AddDataToDB) close() {
	fmt.Println("Close Database.")
}

// Client code
func main() {
	o := &AddDataToDB{}
	o.addData()
}

/*
Open Database.
Add data to Database.
Close Database.
*/

File: Design-Patterns-Using-Cpp/Template/TemplatePatternData.cpp
#include <iostream>

// AddDataTemplate class
class AddDataTemplate {
public:
    void addData() {
        open();
        add();
        close();
    }

    virtual ~AddDataTemplate() {}

private:
    virtual void open() = 0;
    virtual void add() = 0;
    virtual void close() = 0;
};

// AddDataToFile class
class AddDataToFile : public AddDataTemplate {
private:
    void open() override {
        std::cout << "Open file." << std::endl;
    }

    void add() override {
        std::cout << "Add data to file." << std::endl;
    }

    void close() override {
        std::cout << "Close file." << std::endl;
    }
};

// AddDataToDB class
class AddDataToDB : public AddDataTemplate {
private:
    void open() override {
        std::cout << "Open Database." << std::endl;
    }

    void add() override {
        std::cout << "Add data to Database." << std::endl;
    }

    void close() override {
        std::cout << "Close Database." << std::endl;
    }
};

// Client code
int main() {
    AddDataTemplate* addDataToDB = new AddDataToDB();
    addDataToDB->addData();

    delete addDataToDB; // Freeing allocated memory

    return 0;
}

/*
Open Database.
Add data to Database.
Close Database.
*/